### PR TITLE
Motor Control

### PR DESCRIPTION
**Pull request by LandryNorris**
Add command to set motor power to sample.

### CODE REVIEW

**Comment by LandryNorris** [comment]
Depends on #29

File: packages/sample/src/main.ts [hunk 1]
```diff
@@ -18,6 +20,18 @@ if(options.list) {
...
+if(options.motor) {
+    let power: number = Number(options.motor);
+
+    const hubs: ExpansionHub[] = await getConnectedExpansionHubs();
+    let hub = hubs[0]
+
+    await hub.setMotorChannelMode(1, 0, true);
+    console.log(`Power is ${power}, type is ${typeof power}`)
+    await hub.setMotorConstantPower(1, power);
+    await hub.setMotorChannelEnable(1, true);
```
Comment (NoahAndrews) [inline packages/sample/src/main.ts:32]: Methods used in the sample must be documented. Also, they should accept values from -1 to 1, not whatever it is now.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/sample/src/main.ts (hunk 1 above)
Comment (LandryNorris) [reply packages/sample/src/main.ts]: Updated it to [-1,1]

Comment (LandryNorris) [reply packages/sample/src/main.ts]: Added docs

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 2]
```diff
@@ -67,13 +67,31 @@ export interface ExpansionHub extends RevHub {
     getI2CReadStatus(i2cChannel: number): Promise<I2CReadStatus>;
 
     // Motor
+    /**
+     * Configure a specific motor
+     * @param motorChannel
+     * @param motorMode 0 for constant power, 1 for constant velocity, 2 for position
+     * @param floatAtZero whether to coast or break when power is set to 0
+     */
     setMotorChannelMode(motorChannel: number, motorMode: number, floatAtZero: boolean): Promise<void>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:76]: `motorMode` should be typed as an enum for the high-level library

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 3]
```diff
@@ -67,13 +67,31 @@ export interface ExpansionHub extends RevHub {
     getI2CReadStatus(i2cChannel: number): Promise<I2CReadStatus>;
 
     // Motor
+    /**
+     * Configure a specific motor
+     * @param motorChannel
+     * @param motorMode 0 for constant power, 1 for constant velocity, 2 for position
+     * @param floatAtZero whether to coast or break when power is set to 0
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:74]:
```suggestion
     * @param floatAtZero whether to coast when power is set to 0. If this is set to true, the motor will brake at 0 instead.
```

File: packages/sample/src/main.ts [hunk 4]
```diff
@@ -6,6 +6,9 @@ const program = new Command();
 
 program.version('1.0.0')
     .option('-l --list', 'List connected devices')
+    .option('-c --count <counts>', 'Set encoder count.')
+    .option('-m --motor <index>', 'Set motor index. Must also include -p or -c')
+    .option('-p --power <power>', 'Set motor power')
```
Comment (NoahAndrews) [inline packages/sample/src/main.ts:11]: `-c` and especially `-p` should only work if `-m` is specified. You don't know what the motors in question are hooked up to, and you don't want the wrong thing to start spinning if they forget to specify which motor. Also, IMO the `-m` description should reference the long forms of options, not the short forms.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 2 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts]: Implemented the enum

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 5]
```diff
@@ -96,14 +97,27 @@ export interface ExpansionHub extends RevHub {
...
+    ): Promise<{ motorMode: MotorMode; floatAtZero: boolean }>;
+
+    /**
+     * Enable a motor. The motor must be configured before enabling.
+     * @param motorChannel
+     * @param enable
+     */
+
     setMotorChannelEnable(motorChannel: number, enable: boolean): Promise<void>;
     getMotorChannelEnable(motorChannel: number): Promise<boolean>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:122]: This and the other motor-related getters need documentation

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 5 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts:133]: Documented

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 6]
```diff
@@ -96,44 +97,123 @@ export interface ExpansionHub extends RevHub {
...
+     * @param motorChannel
+     */
     getMotorChannelEnable(motorChannel: number): Promise<boolean>;
     setMotorChannelCurrentAlertLevel(
         motorChannel: number,
         currentLimit_mA: number,
     ): Promise<void>;
+
+    /**
+     * Get the present current alert (mA)
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:140]:
```suggestion
     * Get the current level in mA that will trigger the current alert for a given motor channel
```

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 7]
```diff
@@ -96,44 +97,123 @@ export interface ExpansionHub extends RevHub {
...
+     */
+
     setMotorChannelEnable(motorChannel: number, enable: boolean): Promise<void>;
+
+    /**
+     * Check whether a motor channel is currently enabled
+     * @param motorChannel
+     */
     getMotorChannelEnable(motorChannel: number): Promise<boolean>;
     setMotorChannelCurrentAlertLevel(
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:134]: Needs documentation

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 8]
```diff
@@ -96,44 +97,123 @@ export interface ExpansionHub extends RevHub {
...
+    /**
+     * Get the current target position for a motor
+     * @param motorChannel
+     */
     getMotorTargetPosition(
         motorChannel: number,
     ): Promise<{ targetPosition: number; targetTolerance: number }>;
+
+    /**
+     * Check if a motor has reached its target
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:186]: Many of these methods need documentation that explains how they interact with the current motor mode. For example, this method is only applicable to RUN_TO_POSITION mode.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 7 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts:140]: Added documentation

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 8 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts:210]: Added note where applicable

**Comment by LandryNorris** [comment]
Added encoder sub-commands for motor and added flag to reset encoder counts

File: packages/expansion-hub/src/MotorMode.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+//The order of these should match the RHSPlib documentation. We rely on the
+//numerical value matching.
+export enum MotorMode {
+    CONSTANT_POWER,
+    CONSTANT_VELOCITY,
+    TARGET_POSITION,
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/MotorMode.ts:6]: The SDK uses horrible names for constant power and constant velocity currently, and there was discussion of what we should rename them to here: https://github.com/FIRST-Tech-Challenge/ftc_sdk/issues/1990

I don't love `TARGET_POSITION`. I think I do like Chris's suggestion of `UNREGULATED`, `REGULATED_VELOCITY`, `REGULATED_POSITION`. "Constant" implies that it does a perfect job of maintaining the desired value.

Comment (LandryNorris) [reply packages/expansion-hub/src/MotorMode.ts:6]: How does OPEN_LOOP sound as a replacement for CONSTANT_POWER? This isn't the FTC SDK, so we don't have the same assumptions about user knowledge.

Comment (NoahAndrews) [reply packages/expansion-hub/src/MotorMode.ts:6]: I'm fine with that.

File: packages/core/src/MotorMode.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+//The order of these should match the RHSPlib documentation. We rely on the
+//numerical value matching.
+export enum MotorMode {
+    /**
+     * This mode controls the motor's speed by setting the pwm duty cycle.
```
Comment (NoahAndrews) [inline packages/core/src/MotorMode.ts:5]:
```suggestion
     * This mode controls the motor's speed by setting the PWM duty cycle.
```

File: packages/sample/src/command/motor.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+export async function getMotorPid(hub: ExpansionHub, channel: number) {
+    let pid: PidCoefficients = await hub.getMotorPIDCoefficients(
+        channel,
+        MotorMode.REGULATED_VELOCITY,
+    );
+
+    console.log(JSON.stringify(pid));
+}
```
Comment (NoahAndrews) [inline packages/sample/src/command/motor.ts:109]: There doesn't have to be both versions right now, but the sample needs to make it clear whether you're setting the velocity PID or the position PID.

File: packages/sample/src/main.ts [hunk 12]
```diff
@@ -82,34 +121,166 @@ motorCommand
...
+    .description("Set PID coefficients for a motor")
+    .action(async (channel) => {
+        let channelNumber = Number(channel);
+        let hubs = await openConnectedExpansionHubs();
+        let hub = hubs[0];
+
+        runOnSigint(() => {
+            hub.close();
+        });
+
```
Comment (NoahAndrews) [inline packages/sample/src/main.ts:179]: Velocity vs position PID needs to be clearer.

**Review by NoahAndrews — DISMISSED** [review]

File: packages/sample/src/command/motor.ts (hunk 11 above)
Comment (NoahAndrews) [reply packages/sample/src/command/motor.ts:109]: The names of these functions should reflect that they only deal with velocity PID.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

**Pull request merged**